src/shims/io.rs

```diff
@@ -0,0 +1,170 @@
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let cmd = this.read_scalar(cmd_op)?.to_i32()?;
+
+        if cmd == this.eval_libc_i32("F_SETFD")? {
+            // This does not affect the file itself. Certain flags might require changing the file
+            // or the way it is accessed somehow.
+            let flag = this.read_scalar(arg_op.unwrap())?.to_i32()?;
+            // The only usage of this in stdlib at the moment is to enable the `FD_CLOEXEC` flag.
+            let fd_cloexec = this.eval_libc_i32("FD_CLOEXEC")?;
+            if let Some(old_flag) = this.machine.file_handler.flags.get_mut(&fd) {
+                if flag ^ *old_flag == fd_cloexec {
+                    *old_flag = flag;
+                } else {
+                    throw_unsup_format!("Unsupported arg {:#x} for `F_SETFD`", flag);
+                }
+            }
+            Ok(0)
+        } else if cmd == this.eval_libc_i32("F_GETFD")? {
+            if let Some(flag) = this.machine.file_handler.flags.get(&fd) {
+                Ok(*flag)
+            } else {
+                this.machine.last_error = this.eval_libc_i32("EBADF")? as u32;
+                Ok(-1)
+            }
+        } else {
+            throw_unsup_format!("Unsupported command {:#x}", cmd);
+        }
+    }
+
+    fn close(&mut self, fd_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+
+        if let Some(file) = this.machine.file_handler.files.remove(&fd) {
+            match file.sync_all() {
+                Ok(()) => Ok(0),
+                Err(e) => {
+                    this.machine.last_error = e.raw_os_error().unwrap() as u32;
+                    Ok(-1)
+                }
+            }
+        } else {
+            this.machine.last_error = this.eval_libc_i32("EBADF")? as u32;
+            Ok(-1)
+        }
+    }
+
+    fn read(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        buf_op: OpTy<'tcx, Tag>,
+        count_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i64> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let tcx = &{ this.tcx.tcx };
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
+        let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
+
+        if let Some(file) = this.machine.file_handler.files.get_mut(&fd) {
+            let mut bytes = vec![0; count as usize];
```

Comment:
I'm not sure if that works out, but I think you can use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/struct.Allocation.html#method.get_bytes_mut to get a `&mut [u8]` to the destination directly and then read directly into the destination allocation.

that would probably require removing the `FileHandle` from the `files` map and inserting it back in later, because otherwise you borrow `this` via `memory_mut()` and via `this.machine`